File: 프로그래머스/2/131704. 택배상자/택배상자.java
import java.util.*;

class Solution {
    public int solution(int[] order) {
        int answer = 0;
        int len = order.length;
        
        Deque<Integer> subContainer = new ArrayDeque<>();
        int i = 1;
        
        for(int ord: order){
            if((i != ord && i > len) && (!subContainer.isEmpty() && ord != subContainer.peekLast())){
                break;
            }
            
            while(true){
                if(ord == i){
                    // 바로 트럭에 싣는 경우
                    answer++;
                    i++;
                    break;
                } else if (!subContainer.isEmpty() && ord == subContainer.peekLast()){
                    // 보조 컨테이너에서 싣는 경우
                    subContainer.pollLast();
                    answer++;
                    break;
                } else if (i<=len) {
                    // 순서가 맞지 않아 보조 컨테이너에 보관하는 경우
                    subContainer.offerLast(i);
                    i++;
                } else {
                    // 방법이 없는 경우
                    break;
                }     
            }
        }
        
        
        return answer;
    }
}

File: 프로그래머스/2/131704. 택배상자/택배상자.py
def solution(order):
    answer = 0
    idx = 1
    l = len(order)
    sub_container = [] ## stack
    
    for o in order:
        if (idx > l and idx != o) and (sub_container and sub_container[-1] != o):
            ## 방법이 없는 경우
            break
            
        while True:
            if o == idx:
                ## 바로 트럭에 싣는 경우
                answer += 1
                idx += 1
                break
            elif (sub_container and sub_container[-1] == o):
                ## 보조 컨테이너 벨트에서 싣는 경우
                sub_container.pop()
                answer += 1
                break
            elif idx <= l:
                ## 보조 컨테이너 벨트에 보관하는 경우
                sub_container.append(idx)
                idx += 1
            else:
                break
    
    return answer